app/hooks/use-device-location.ts

```diff
@@ -0,0 +1,47 @@
+import { useApolloClient } from "@apollo/client"
+import { useEffect, useState } from "react"
+import { updateCountryCode } from "@app/graphql/client-only-query"
+import { useCountryCodeQuery } from "@app/graphql/generated"
+import axios from "axios"
+import { CountryCode } from "libphonenumber-js/mobile"
+
+const useDeviceLocation = () => {
+  const client = useApolloClient()
+  const { data } = useCountryCodeQuery()
+
+  const [loading, setLoading] = useState(true)
+  const [countryCode, setCountryCode] = useState<CountryCode>("SV")
+
+  useEffect(() => {
+    if (data) {
```

Comment:
are you waiting for `data` for the throw case?

Reply:
an alternative that would have less nesting would be to split the useEffect in 2:
- one useEffect that `setCountryCode` on data when its available (almost immediately)
- one useEffect that goes through `https://ipapi.co/json/` and `setCountryCode` again when the code is resolved. there is no longer a need to have the `catch` branch here because the value would already have been set.

functionality that should be the same but that may make code easier to read

Reply:
I've put the two cases (useCountryCodeQuery and iPapi) together because they both need to finish in order for me to change **loading** to false. This is to indicate that the **initialLocation* has been established and the Map can go ahead and mount. I think having the two of them together shows that they're coupled.

As per the data throw case I agree that for good form, I should account for the error as well. I'll add another useEffect for that that'll setLoading to false and it'll allow the Map to render at the default 'SV' country code.

Reply:
For the useCountryCodeQuery I added an error case in a useEffect in case it throws. 

As per the dividing of code here, I think they should stay together because technically it's a race condition. useCountryCodeQuery would never realistically finish second, but nonetheless it's async and should be treated as possible. And I think it's best to show that all this is executing in the same order because I can only toggle **loading** to false when both are finished.